Logs backend MVP

jazeved0 opened the pull request:
## Summary

This pull request is designed to be the MVP for the backend event pipeline & GraphQL API of the logs service.

### Goals

- Have all services running to ingest a small subset of log events into a data store. These events are, at the moment:
  - MessageSend
  - MessageReply
  - ReactionAdd
  - ReactionRemove
  - ReactionBulkRemove
  - MemberJoin
  - MemberLeave
- Provide a simple GraphQL API for querying the log data
- Make log ingestion opt-in via the use of feature flags

### Non-goals

- Authentication/rate-limiting for the GraphQL API -- This API will not be publicly exposed to the internet and will only be usable from within the cluster for testing. Authentication/rate-limiting will be added at a later point
- Full coverage of all possible log events
- Inclusion of a history-scanning service -- this is something that will be added before product launch that will make it possible to replicate Discord audit log events into our own log events in an intelligent way that knows what periods in time it has already submitted events for.
- Nickname/color revision tracking -- this is a feature that will eventually be present for use in the web UI to show nicknames and colors in their actual point-in-time representation by tracking when they change.
- Guild uptime tracking -- this is a monitoring feature that will be present and track when the gateway ingress service is online and receiving events from Discord. It will inform the history-scanning service for potentially recovering certain events that were lost, and might also expose an API for viewing uptime history on the dashboard
- Any UI work -- the API is intended to be usable only via GraphQL clients, including a self-served "GraphQL playground" web app
- Log retention work -- we plan on implementing log retention policies and then making longer retention periods a premium feature.
- Internal authentication -- I don't have the expertise for this so I'm putting it off until after the backend MVP, but it is something we should absolutely implement before release.
- Configuration of Elasticsearch -- it's likely that we will have to do some configuration of Elasticsearch's refreshing behavior in order to strike a balance between indexing speed and data freshness. [Discord used a lazy refreshing behavior](https://blog.discord.com/how-discord-indexes-billions-of-messages-e3d5e9be866f#0059) using redis which sounds straightforward to implement, but would add a dependency on Redis that is currently not being used anywhere. More thought should be put into this at a later point in time.
- Feature-complete filtering/sorting in the GraphQL schema including full-text search
- Finalizing the GraphQL schema to allow ergonomic queries with entity variants

## Testing

To test the MVP, the following steps can be run:

1. Create the `secrets.yaml` file and populate it as needed. Of special note are all the environment variables that start with `LOGS_` and don't have a value. These include:
    - `LOGS_GATEWAY_INGRESS_CONFIG_SECRETS__DISCORD_TOKEN` and `LOGS_GATEWAY_NORMALIZE_CONFIG_SECRETS__DISCORD_TOKEN`, which both need to be populated with a valid Discord bot token
2. `tilt up -- --no-core --enable=logs` in the root of the repository
3. Have your test bot be in a guild with your Discord user for testing. Then, copy the ID of the guild
4. Use `psql` to enable the feature flag for your guild:
    1. Run `psql -h localhost -p 5432 -U autbot`, and then type the password (`autism` by default)
    2. In the Postgres REPL, run:
        1. `INSERT INTO tb_feature_flags (name,open) VALUES ('beta/logs-indexing',false);`
        2. `INSERT INTO tb_guild_features (guild_id,feature_id) VALUES (596436041045639179,1);` Be sure to replace the guild ID in this command with your own
5. Wait some duration for the feature flag to propagate. This should not take longer than a minute or two.
7. Perform one of the supported actions in your guild in a channel that the bot can see, such as adding a reaction to a message
8. Open up the GraphQL playground at http://localhost:8174/playground and execute the following query:<br /><br />
    ```graphql
    query DemoQuery {
        allEvent {
            snapshot
            pageInfo {
              currentPage
              hasPreviousPage
              hasNextPage
              itemCount
              pageCount
              perPage
              totalCount
            }
            nodes {
                id
                timestamp
                type
                guildId
                channelId
                agentId
                agentType
                agentSpecialType
                subjectId
                subjectType
                auxiliaryId
                auxiliaryType
                content
            }
        }
    }
    ```
    The response should look something like:<br /><br />
    ```json
    {
      "data": {
        "allEvent": {
          "snapshot": "bGdldl8xdnFVQzlCRVdQUmI3Vm9ZZFRJVU9abFBpRVo=",
          "pageInfo": {
            "currentPage": 0,
            "hasPreviousPage": false,
            "hasNextPage": false,
            "itemCount": 4,
            "pageCount": 1,
            "perPage": 100,
            "totalCount": 4
          },
          "nodes": [
            {
              "id": "lgev_1vkjznmHsDYYJFeB9y0CD9eCohr",
              "timestamp": "1627113245056",
              "type": "REACTION_ADD",
              "guildId": "607637793107345431",
              "channelId": "608075539629998081",
              "agentId": "74646589213257728",
              "agentType": "USER_LIKE",
              "agentSpecialType": "DEFAULT",
              "subjectId": "867594760281653259",
              "subjectType": "MESSAGE",
              "auxiliaryId": null,
              "auxiliaryType": "NONE",
              "content": "😎 :sunglasses:"
            }
          ]
        }
      }
    }
    ```

jazeved0 committed:
feat: add basic gateway event listening

jazeved0 committed:
chore: add new crate for shared lib code

jazeved0 committed:
feat(logs): start to normalize gateway events

jazeved0 committed:
fix(logs): assign correct ids for IdProvisioner

jazeved0 committed:
feat(id): remove extra bitwise op

jazeved0 committed:
feat(logs): add fully implemented normalization for channel creation

jazeved0 committed:
fix(logs): reduce nesting; refactor

jazeved0 committed:
fix(logs): update comments

jazeved0 committed:
feat(logs): add channel delete hybrid logging

jazeved0 committed:
feat: generalize channel events

jazeved0 committed:
fix(logs): use borrows of Channel instead of cloning

jazeved0 committed:
feat(logs): harden gateway ingress

jazeved0 committed:
feat: use growing interval search strategy

jammiess commented on logs/ingress/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+lazy_static = "1.4.0"
+log = "0.4.8"
+env_logger = "0.7.1"
+tokio = { version = "0.2.20", features = ["macros"] }
+futures = "0.3.4"
+logs_lib = { path = "../lib" }
+serde = { version = "1.0.111", features = ["derive"] }
+serde_json = "1.0.55"
+backoff = "0.1.6"
+reqwest = "0.10.6"
```
Unless it's necessary to have these exact versions, you can just do X.Y and if the library writers do semantic versioning properly, you'll get the newest version that doesn't break the API.

jazeved0 committed:
fix(logs): reduce timestamp syscalls

jammiess commented on logs/ingress/src/audit_log.rs:
```diff
@@ -0,0 +1,252 @@
+use serenity::http::Http;
+use serenity::model::guild::AuditLogEntry;
+use serenity::model::id::{AuditLogEntryId, GuildId, UserId};
+use std::sync::Arc;
```
If you're using this for async stuff, Tokio implements it's own async safe reference types that are probably better to use.

jammiess commented on logs/ingress/src/audit_log.rs:
```diff
@@ -0,0 +1,252 @@
+        return match &**boxed_err {
+            serenity::http::HttpError::UnsuccessfulRequest(response) => {
+                if response.status_code == StatusCode::UNAUTHORIZED {
+                    return Some(response.clone());
+                } else {
+                    return None;
```
you don't need these returns as you're returning the match block. just have the if and else evaluate to the Some(_) or None

jammiess commented on logs/ingress/src/audit_log.rs:
```diff
@@ -0,0 +1,252 @@
+        // attempt to find the desired matching entry
+        let mut oldest: Option<AuditLogEntryId> = None;
+        for (key, value) in entries {
+            if (search.matches)(&value) && search.strategy.matches(timing, &value, search) {
```
You don't need the parens around the function pointer.

jazeved0 replied:
rustc yells at me that `matches` is a `field, not a method` :(

jammiess replied:
Weird. Not sure why it doesn't work.

jammiess replied:
You could maybe try a function pointer instead for your variadic type. Have `P: fn(&input) -> bool`

jazeved0 replied:
Wouldn't that mean that I couldn't pass closures to the function? I'd rather keep the callsite interface as ergonomic as possible.

jammiess replied:
Rust will autocast a closure to a function pointer. But this is mainly me just finding that syntax ugly and not an actual problem with the code so don't worry about it too much.

jammiess commented on logs/lib/src/time.rs:
```diff
@@ -0,0 +1,25 @@
+    /// Invokes `clock_gettime` from time.h in libc to get a `timespec` struct
+    fn get_time() -> timespec {
+        let mut tp: timespec = unsafe { mem::zeroed() };
+        unsafe {
+            clock_gettime(CLOCK_REALTIME, &mut tp);
+        }
```
It's best practice to have SAFETY: reasoning comments for every unsafe block in rust that explains your reasoning for why something is safe

jazeved0 replied:
it's an FFI call to a `libc` function, but I'll add a comment

jazeved0 committed:
fix(logs): incorporate feedback; improve syntax

jazeved0 committed:
fix(logs): correct error message

jazeved0 committed:
feat(logs): extract audit log upgrading

jazeved0 committed:
chore: remove lock files

jazeved0 committed:
feat: start working on data processor

jazeved0 committed:
feat: re-write ingress entrypoint to use twilight

jazeved0 committed:
feat: add basic stream workflow

jazeved0 committed:
feat: add working pipeline for message send events

jazeved0 committed:
feat: work on adding protobuf definitions + logging service

jazeved0 committed:
feat: add gRPC client calls in ingress

jazeved0 committed:
chore: fix lint errors

jazeved0 committed:
feat(logs): add basic logging gRPC server

jazeved0 committed:
feat(logs) add working gRPC client-server setup

TODO: add logging service functionality + ES database

jazeved0 committed:
feat: add Dockerfiles

jazeved0 committed:
feat(logs): add basic import functionality

jazeved0 committed:
feat(logs): modify schema

jazeved0 committed:
fix(logs): move lib folder in Dockerfiles

jazeved0 committed:
feat(logs): add search gRPC call

jazeved0 committed:
chore(logs): start adding juniper integration

jazeved0 committed:
feat(logs): add basic GraphQL API

jazeved0 committed:
feat(logs): add pagination API with opaque snapshot tokens

jazeved0 committed:
fix(logs): rename GraphQL type StoredEvent to Event

jazeved0 committed:
feat(logs): add gRPC search API

jazeved0 committed:
fix(logs): config & page info calculations

jazeved0 committed:
feat(logs): add basic filtering via GraphQL

jazeved0 committed:
feat(logs): add stateful uptime tracking to gateway-ingress

jazeved0 committed:
feat(logs): add framework for tracking active guilds

jazeved0 committed:
feat(logs): add impl for event type forwarding

jazeved0 committed:
feat(logs): add manual active status retrieval

TODO: add stream flat-map'ing and polling behavior

jazeved0 committed:
feat(logs): add basic polling behavior

TODO: add ability to intercept events; re-factor API to allow for mpsc channel + consuming futures instead of piped stream

jazeved0 committed:
fix(logs): update config schema/example config

jazeved0 committed:
feat(logs): add uptime event filtering for Online events

TODO: add support for Offline & Heartbeat events

jazeved0 committed:
feat(logs): finish initial impl of ActiveGuilds

jazeved0 committed:
feat(logs): add basic uptime protobuf definitions

jazeved0 committed:
feat(logs): work on refactor/dockerization of gateway-ingress

jazeved0 committed:
chore(logs): clean up crate::active_guilds

jazeved0 committed:
feat(logs): work on docker-compose deployment

jazeved0 committed:
feat(logs): add successful deployment config

jazeved0 committed:
feat(logs): add random session to uptime events

jazeved0 committed:
fix(feature-gate): correct DB url & add connection backoff

Also adds .gitignores's; TODO test

jazeved0 committed:
fix(feature-gate): remove build artifact

johnyburd committed:
works

johnyburd committed:
fix some problems (this commit got clobbered by rebase)

jazeved0 committed:
feat(logs): work on adding normalize queue consumer

jazeved0 committed:
feat(logs): add submission service stub

jazeved0 committed:
feat(logs): update submission to use revision protobuf

jazeved0 committed:
feat(logs): overhaul normalization data pipeline

jazeved0 committed:
feat(logs): use quorum queue; move declaration to producer

jazeved0 committed:
chore(logs): bump RabbitMQ version to 3.8

jazeved0 committed:
feat(logs): finish channel create procesor

jazeved0 committed:
feat(logs): add chain, start adding other channel events

jazeved0 committed:
feat(logs): work on dividng events into hybrid/gateway/audit log

jazeved0 committed:
feat: add emoji shortcode manipulation for reaction event formatting

johnyburd committed:
update for sandbox & fix ignores (this got clobbered by rebase)

jazeved0 committed:
feat: rename example.yaml file (this commit got clobbered by rebase)

jazeved0 committed:
feat: add minimal deployment for feature-gate on Tilt/k8s

jazeved0 committed:
feat: insane crazy juicer hot reload tilt k8s rust deployment

jazeved0 committed:
feat: add optional component support to tiltfile

Additionally, adds flag to enable hot-reloading of Rust services (requires Cargo installed locally)

jazeved0 committed:
feat: add docker-compose for building in deployment directory

jazeved0 committed:
feat: add production kubernetes templates

jazeved0 committed:
chore: remove deployment config from repo

jazeved0 committed:
feat(logs): add submission-logstash service/config, refactor protos

jazeved0 committed:
feat(logs): update tokio, add submission implementation

jazeved0 committed:
add successful ingress into ES

jazeved0 committed:
feat(logs): add working impl of MVP

jazeved0 committed:
feat(logs): remove logstash, use batching submission

jammiess commented on .editorconfig:
```diff
@@ -0,0 +1,15 @@
+root = true
```
@johnyburd Do we really need this file?

johnyburd replied:
yes

jammiess commented on lib/amqp-pool-rs/Cargo.lock:
```diff
@@ -0,0 +1,919 @@
+# This file is automatically @generated by Cargo.
```
This is in a library? Shouldn't have a lock file

jammiess commented on lib/config-backoff-rs/Cargo.lock:
```diff
@@ -0,0 +1,187 @@
+# This file is automatically @generated by Cargo.
```
No locks in a library

jammiess commented on lib/id-rs/Cargo.lock:
```diff
@@ -0,0 +1,133 @@
+# This file is automatically @generated by Cargo.
```
No lock in library

jammiess commented on lib/id-rs/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+    pub fn with_ts(&self, timestamp: u64) -> HoarFrost {
+        // Note: we can use Ordering::Relaxed here because the overall ordering
+        // doesn't really matter; all that matters is that the operation is atomic
+        // (since the timestamp is at a more significant bit than the counter)
```
Good comment

jazeved0 committed:
feat(logs): polish logs/submission service

jammiess commented on lib/id-rs/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+
+            // unsafe needed for FFI call to libc
+            // (it's (almost?) impossible for this call to break safety)
+            let result = unsafe { clock_gettime(CLOCK_REALTIME, &mut tp_out) };
```
I would recommend swapping to using the rust stdlib for this. It's just a lot more idiomatic and there's no safety worries. Also, does the underlying time need to be monotonic or not? Cause the libc sycall will not make that guarantee.

jazeved0 committed:
feat(logs): add logging to id library

jazeved0 committed:
chore(logs): remove Cargo lock for library

jazeved0 committed:
chore(logs): add lockfiles to library gitignores

jazeved0 committed:
feat(logs): add config for running services in Tilt

jazeved0 committed:
feat: work on organization/consistency

these changes are WIP

johnyburd committed:
untested http routes for logs search

jazeved0 committed:
feat: work on making things less terrible

jazeved0 committed:
feat: work on rewriting logsgateway-ingress after twilight update

jazeved0 committed:
feat: make gateway-normalize use slog

jazeved0 committed:
feat: work on cleaning up services but they don't work

jazeved0 committed:
feat: i made the tests pass but I forgot why i wrote the code in the first place

jammiess commented on lib/id-rs/src/lib.rs:
```diff
@@ -0,0 +1,165 @@
+pub struct IdProvisioner {
+    combined_process_id: u64,
+    internal_counter: AtomicU64,
+    logger: Option<Logger>,
```
The logger shouldn't be an option if it is always going to be used

jazeved0 committed:
fix: always use loggers with id provisioners

jammiess commented on logs/gateway-ingress/src/publish.rs:
```diff
@@ -0,0 +1,852 @@
+        }
+
+        // Print out the raw bytes for debugging
+        slog::debug!(
+            logger,
+            "raw event bytes";
+            "event_bytes" => base64::encode(&raw_bytes),
+        );
```
Do we need debug info for an event from a potentially inactive guild?

jazeved0 replied:
It's only used when the log level is debug, so it doesn't do anything in production. Because when developing locally the bot won't be in many guilds, this shouldn't be very noisy (and might be helpful in debugging events being dropped due to buggy code flagging them as unnecessary).

jazeved0 committed:
chore: prepend anyhow:: before Result<> values to be clearer

jammiess commented on logs/gateway-ingress/src/publish.rs:
```diff
@@ -0,0 +1,852 @@
+        if let serde_json::Value::Object(map) = value {
+            // Attempt to find the ".d" value (contains the Gateway message payload)
+            // https://discord.com/developers/docs/topics/gateway#payloads-gateway-payload-structure
+            let mut map = map;
```
Wouldn't it be better to put the `mut` in the let statement?

jazeved0 committed:
fix: remove Cargo.lock copies for libraries

jammiess commented on logs/gateway-normalize/src/gateway/processors.rs:
```diff
@@ -0,0 +1,487 @@
+mod message {
+    use crate::gateway::ProcessorFleet;
+
+    pub fn register_all(_fleet: &mut ProcessorFleet) {
+        // TODO implement MessageSend processor
+        // TODO implement MessageReply processor
+        // TODO implement MessageEdit processor
+        // TODO implement MessageDelete processor
+        // TODO implement MessageBulkDelete processor
+    }
```
Does this need to be implemented?

jazeved0 replied:
Nope :) See the non-goals

jammiess commented on logs/gateway-normalize/src/gateway/processors.rs:
```diff
@@ -0,0 +1,487 @@
+mod interaction {
+    use crate::gateway::ProcessorFleet;
+
+    pub fn register_all(_fleet: &mut ProcessorFleet) {
+        // TODO implement InteractionCreate processor
+    }
```
Should this be implemented?

jazeved0 replied:
Nope :)

jammiess commented on logs/search/src/graphql/inputs.rs:
```diff
@@ -0,0 +1,606 @@
+    auxiliary_id: Option<UIntStringOptionFilterInput>,
+    auxiliary_type: Option<EnumFilterInput>,
+    content: Option<TextOptionFilterInput>,
+    // TODO implement content metadata filter
```
You left a lot of todos in your PR

jazeved0 replied:
That's true :)

jazeved0 committed:
fix: add needed files to docker contexts

jazeved0 committed:
feat: use ARG's instead of ENV's; share between stages

jazeved0 committed:
fix: add schema to logs-submission container

jazeved0 committed:
feat: use safer id generation

jazeved0 committed:
fix: move mut into pattern to be more terse

jazeved0 committed:
feat: use rocket instead of hyper; fix API frontend

jazeved0 committed:
chore: move single-file directory module to single file

jazeved0 committed:
fix: add back in fairings code

jazeved0 committed:
feat: use opaque encoding for snapshot token

jazeved0 committed:
chore: fix clippy warnings

jazeved0 committed:
work a bit on new submission logic

jazeved0 committed:
do some rewriting (in rust)

jazeved0 committed:
remove old code

jazeved0 committed:
restructure some functions to not be 200 lines long

jazeved0 committed:
we do some doc writing

jazeved0 committed:
commit wip changes

jazeved0 committed:
cut some old code

jazeved0 committed:
make stuff compile again

jazeved0 committed:
clean up some more code

jazeved0 committed:
remove uptime; greatly simplify ingress implementation

jazeved0 committed:
clean up logs-ingress

jazeved0 committed:
rename connect, run cargo fmt

jazeved0 committed:
remove usages of architus_id where needed

jazeved0 committed:
clean up some various things

jazeved0 committed:
add implementation for message send & message reply

jazeved0 committed:
add ES authentication support to logs-submission

jazeved0 commented:
### For logs/submission:
- [x] resolve merge conflicts
- [x] clean up normalize event, move into shared library (logs-lib), including rich content parsing code
- [x] add timeouts to all rpc's
- [x] clean up config impl/templates
- [x] check unused cargo dependencies, ensure all are updated
- [x] clean up build.rs
- [x] clean up tiltfile to be easier to edit
- [x] add doc comments for most functions as needed
- [x] run clippy
- [x] run spell check
- [x] create separate PR containing the following components:
  - logs/submission
  - lib/logs-event-rs
  - elasticsearch
- [ ] create pr for prod deployment config
  - need elasticsearch PV support
  - add elasticsearch, logs/submission as services

### For other services
- [x] add implementation for:
  - [x] message edits 
  - message delete (defer until communication with shard's message cache and/or logs store is implemented)
  - message bulk deletes (defer until communication with shard's message cache and/or logs store is implemented)
  - [ ] interaction creates
- [x] clean up logs for initial handle update send
- [ ] figure out why events sometimes get stuck in the gateway queue (normalize isn't working?)
- [ ] fix known emoji edge cases
- [ ] add timeouts to all rpc's
- [ ] clean up config impl/templates
- [ ] check unused cargo dependencies, ensure all are updated
- [ ] clean up tiltfile to be easier to edit
- [ ] add doc comments for most functions as needed
- [ ] run clippy
- [ ] run fmt
- [ ] run spell check
- [ ] create separate PR to upgrade rabbitmq version (to support quorum queues)
- [ ] create 2 separate PRs for:
  - [ ] logs/gateway-ingress logs/gateway/normalize
  - [ ] logs/search api
- [ ] create pr's for prod deployment configs for each component
  - need rabbitmq PV support

jazeved0 committed:
clean up log message

jazeved0 committed:
work on some structure; implementation of message edit

jazeved0 committed:
add logs-lib-rs

jazeved0 committed:
clean up logs-lib, add remaining unit tests

jazeved0 committed:
Merge remote-tracking branch 'origin/develop' into feature/logs-service

jazeved0 committed:
add timeouts to rpc calls in submission

jazeved0 committed:
work on fixing docker stuff, making tiltfile api less bad

jazeved0 committed:
use newer format for all logs services, make init.sh not terrible

jazeved0 committed:
fixes to dockerfiles

jazeved0 committed:
add more details to doc comment

jazeved0 committed:
apply clippy fixes

jazeved0 committed:
Merge remote-tracking branch 'origin/develop' into feature/logs-service

jazeved0 committed:
move all python lib files to lib/python-common

jazeved0 committed:
ensure paths are valid to lib/python-common & lib/proto

jazeved0 committed:
use individual allow-list-based dockerignores

jazeved0 committed:
add sandbox dockerignore

jazeved0 committed:
add docs to tilt functions, clean up

jazeved0 committed:
remove extra change

jazeved0 committed:
Add scoped docker ignores to the .gitattributes file

jazeved0 committed:
Merge branch 'jazevedo620/overhaul-docker-builds' into feature/logs-service

jazeved0 committed:
remove additional files from branch

jazeved0 committed:
run clippy & fmt
